tests/AggregationBlogPost/article.md

```diff
@@ -0,0 +1,713 @@
+helps you write a pipeline, but it doesn't really make existing pipelines more
+maintainable. Yes, it makes them easier to read, but a complex pipeline will
+remain just as complex. So, let's discuss some refactorings we can make to make
+the aggregation pipeline easier to read, but also to make parts of the pipeline
+reusable. Note that all of these example apply the same way to pipelines written
+as PHP arrays, but I'll use the aggregation builder in the example.
+
+Let's look at the first `$group` stage in the original example:
+
+```php
+Stage::group(
+    _id: object(
+        year: Expression::year(Expression::dateFieldPath('reportDate')),
+        month: Expression::month(Expression::dateFieldPath('reportDate')),
+        fuelType: Expression::fieldPath('fuelType'),
+        brand: Expression::fieldPath('station.brand'),
+    ),
+    lowest: Accumulator::min(Expression::doubleFieldPath('price')),
+    highest: Accumulator::max(Expression::doubleFieldPath('price')),
+    average: Accumulator::avg(Expression::doubleFieldPath('price')),
+    count: Accumulator::sum(1),
+);
+```
+
+As you can see, we use the `reportDate` and `price` fields multiple times. An
+obvious refactoring would be to extract a variable for this:
+
+```php
+$reportDate = Expression::dateFieldPath('reportDate');
+$price = Expression::doubleFieldPath('price');
+
+Stage::group(
+    _id: object(
+        year: Expression::year($reportDate),
+        month: Expression::month($reportDate),
+        fuelType: Expression::fieldPath('fuelType'),
+        brand: Expression::fieldPath('station.brand'),
+    ),
+    lowest: Accumulator::min($price),
+    highest: Accumulator::max($price),
+    average: Accumulator::avg($price),
+    count: Accumulator::sum(1),
+);
+```
+
+The `fuelType` and `station.brand` fields could be extracted as well. Since
+these are only used once, I didn't do that, but you may want to do so to favour
+consistency.
+
+### Comments Or Methods
+
+In complex pipelines, you'll often find comments explaining what a certain
+pipeline stage or segment does. You should definitely include comments like
+that, but you can also consider extracting parts of a pipeline to your own
+builder method. If you choose a descriptive method name, this can already
+explain what the stage or segment does, without the reader having to see the
+internal workings.
+
+```php
+public static function groupAndComputeStatistics(
+    stdClass $_id,
+    Expression\ResolvesToDouble $price,
+): GroupStage {
+    return Stage::group(
+        _id: $_id,
+        lowest: Accumulator::min($price),
+        highest: Accumulator::max($price),
+        average: Accumulator::avg($price),
+        count: Accumulator::sum(1),
+    );
+}
+```
+
+When extracting logic, consider if the method can be reused elsewhere. In this
+case, by keeping the `_id` field for the `$group` stage along with the price
+field as a parameter, we can reuse this builder method in a different pipeline. 
```

Comment:
In the interest of self-documenting code, you could consider renaming the `$_id` parameter to `$groupBy`. But I think most users of the aggregation framework would know what `_id` is in this context so probably not necessary.

Reply:
Yes, but this is an excellent point. While users of the aggregation framework would know that `_id` is used for grouping, I still consider `groupBy` a better name for the extracted method. I've updated the sample code 👍